policy/policy-authority-data.go

```
@@ -0,0 +1,145 @@
+)
+
+const whitelisted = "whitelist"
+const blacklisted = "blacklist"
+
+// DomainRule ...
+type DomainRule struct {
+	Host string `db:"host"`
+	Type string `db:"type"`
+}
+
+func reverseName(domain string) string {
+	labels := strings.Split(domain, ".")
+	for i, j := 0, len(labels)-1; i < j; i, j = i+1, j-1 {
+		labels[i], labels[j] = labels[j], labels[i]
+	}
+	return strings.Join(labels, ".")
+}
+
+// PolicyAuthorityDatabaseImpl enforces policy decisions based on various rule
+// lists
+type PolicyAuthorityDatabaseImpl struct {
+	log   *blog.AuditLogger
+	dbMap *gorp.DbMap
+}
+
+// NewPolicyAuthorityDatabaseImpl constructs a Policy Authority Database (and
+// creates tables if they are non-existent)
+func NewPolicyAuthorityDatabaseImpl(dbMap *gorp.DbMap) (padb *PolicyAuthorityDatabaseImpl, err error) {
+	logger := blog.GetAuditLogger()
+
+	dbMap.AddTableWithName(DomainRule{}, "ruleList").SetKeys(false, "Host")
+
+	padb = &PolicyAuthorityDatabaseImpl{
+		dbMap: dbMap,
+		log:   logger,
+	}
+
+	return padb, nil
+}
+
+// LoadRules loads the whitelist and blacklist into the database in a transaction
+// deleting any previous content
+func (padb *PolicyAuthorityDatabaseImpl) LoadRules(rules []DomainRule) error {
+	tx, err := padb.dbMap.Begin()
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	_, err = tx.Exec("DELETE FROM ruleList")
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	for _, r := range rules {
+		r.Host = reverseName(r.Host)
+		tx.Insert(&r)
+	}
+
+	err = tx.Commit()
+	return err
+}
+
+// DumpRules retrieves all DomainRules in the database so they can be written to
+// disk
+func (padb *PolicyAuthorityDatabaseImpl) DumpRules() ([]DomainRule, error) {
+	var dR []DomainRule
+	_, err := padb.dbMap.Select(&dR, "SELECT * FROM ruleList")
+	for _, r := range dR {
+		r.Host = reverseName(r.Host)
+	}
+
+	return dR, err
+}
+
+func (padb *PolicyAuthorityDatabaseImpl) checkBlacklist(host string) error {
```

Is there a reason that these are returning error instead of bool?  It's not like you're going to do anything with the errors you get; you're going to fail closed.